StripeCore/StripeCore/Source/Attestation/StripeAttest.swift

```diff
@@ -0,0 +1,278 @@
+    }
+
+    /// Sign an assertion.
+    /// Will create and attest a new device key if needed.
+    @_spi(STP) public func assert() async throws -> Assertion {
+        do {
+            let assertion = try await _assert()
+            let successAnalytic = GenericAnalytic(event: .assertionSucceeded, params: [:])
+            STPAnalyticsClient.sharedClient.log(analytic: successAnalytic, apiClient: apiClient)
+            return assertion
+        } catch {
+            let errorAnalytic = ErrorAnalytic(event: .assertionFailed, error: error)
+            STPAnalyticsClient.sharedClient.log(analytic: errorAnalytic, apiClient: apiClient)
+            throw error
+        }
+    }
+
+    // MARK: Public structs
+
+    /// Contains the signed data and various information used to sign the request.
+    @_spi(STP) public struct Assertion {
+        /// The signed assertion data.
+        @_spi(STP) public var assertionData: Data
+        /// The `identifierForVendor` for the current app/device pair.
+        @_spi(STP) public var deviceID: String
+        /// The App ID (not fully qualified, it's missing the Team ID prefix)
+        @_spi(STP) public var appID: String
+        /// The key ID
+        @_spi(STP) public var keyID: String
+
+        /// A convenience function to return the fields used in an asserted request.
+        @_spi(STP) public var requestFields: [String: String] {
+            return [ "deviceId": deviceID,
+                     "appID": appID,
+                     "keyID": keyID,
+                     "assertionData": assertionData.base64EncodedString(), ]
+        }
+    }
+
+    @_spi(STP) public enum AttestationError: Error {
+        /// Attestation is not supported on this device.
+        case attestationNotSupported
+        /// Device ID is unavailable.
+        case noDeviceID
+        /// App ID is unavailable.
+        case noAppID
+        /// Retried assertion, but it failed.
+        case secondAssertionFailureAfterRetryingAttestation
+        /// Can't attest any more keys today.
+        case attestationRateLimitExceeded
+        /// The challenge couldn't be converted to UTF-8 data.
+        case invalidChallengeData
+    }
+
+    // MARK: - Internal
+
+    // MARK: - Device-local settings
+    enum DefaultsKeys: String {
+        /// The ID of the attestation key stored in the keychain.
+        case keyID = "STPAttestKeyID"
+        /// The last date we generated an attestation key, used for rate limiting.
+        case lastAttestedDate = "STPAttestKeyLastAttestedDate"
+        /// Whether the current keyID key has been attested successfully.
+        case successfullyAttested = "STPAttestKeySuccessfullyAttested"
+    }
+
+    private static var keyID: String? {
+        get {
+            UserDefaults.standard.string(forKey: DefaultsKeys.keyID.rawValue)
+        }
+        set {
+            UserDefaults.standard.set(newValue, forKey: DefaultsKeys.keyID.rawValue)
+        }
+    }
```

Comment:
Out of scope, but makes me wonder if we should pull out our `@UserDefaults` [property wrapper](https://github.com/stripe/stripe-ios/blob/master/Example/FinancialConnections%20Example/FinancialConnections%20Example/Playground/helpers/UserDefaults.swift) to StripeCore 👀 😅 

```swift
@UserDefault(DefaultsKeys.keyID.rawValue, defaultValue: nil)
static var keyID: String?
```

Reply:
Oooh, neat! We should!